### ISSUE TITLE
Add middleware support to Relic

### ISSUE THREAD

**Issue by vlidholt** [serverpod/serverpod#2499]
Add support for middleware to Serverpod's webserver, Relic, which can be used for custom logging, authentication, data validation, error handling, or other logic. The middleware should be able to modify a request and take over a response. We may potentially want to build a bridge for `shelf` middleware in a separate package.

Serverpod's middleware can be an abstract class `Middleware`. It's primary method is `handleRequest`, which is analog to the `handleRequest` method in a `Route`. The difference is that the method in the middleware can return `null`, in which case control over the response is passed on to the next chained middleware or to the `Route`.

If the middleware needs to override the `Request`, it can do so by calling the `modifyRequest` method.

```dart
abstract class Middleware {
  Future<Response?> handleRequest(Session session, Request request);

  void modifyRequest(Request request) {
 ...
 }
}
```

The middlewares can be injected using a `Router` (see #). Either when creating the `Router`:

```dart
var router = Router(
  middleware: [
    MyCustomLogger(),
    MyCustomAuthentication(),
  ],
);
```

Or, when adding the routes:

```
router.get(
  '/',
  MyRoute,
  middleware: [MyMiddleware()],
);
```

**Comment by vlidholt** [comment]
Yes, the thinking is that the `Request` is immutable, but has a `change` method that returns a modified `Request`. This is how `shelf` works.

**Comment by baudee** [comment]
This feature will be only for the web server ? Could be nice to have it as well for the Endpoints

**Comment by vanlooverenkoen** [comment]
I agree with @baudee having it in the endpoints would simplify my endpoints code by A LOT!

**Comment by dkbast** [comment]
This would be super good for setting the CORS headers :)
